### ISSUE TITLE
Client side validation for conditional rule

### ISSUE THREAD

**Issue by martintro** [JeremySkinner/FluentValidation#303]
In my MVC app: As soon as I add a When condition for a rule the rule is removed from client side. Is there any way to "force" add it to client side even when there is a condition?

**Comment by JeremySkinner** [comment]
No, that's not possible I'm afraid....MVC's support for clientside validation is very limited and only supports a very small subset of rules. If a rule has a condition, it can't be run on the client.

**Comment by martintro** [comment]
Yeah I know that. In this case the when condition is only required on server side when validation, since the field is not shown at all on the client when it shouldn't be validated. Would it be possible to include the rule without a condition on client side and with condition on server side?

**Comment by JeremySkinner** [comment]
Not really I'm afraid...you'd need to have 2 separate rules if you wanted
to do that. You could put them in separate rule sets...once which is
invoked for server side validation and one hats only used for clientside
metadata, but I think you'd then need to duplicate your other rules too.

**Comment by martintro** [comment]
I see. How would that look like? How do I decide which rule sets that should be used server side and client side?

Where in FluentValidation do you filter all rules with conditions to not be included to client side?

**Comment by JeremySkinner** [comment]
So you'd do something like this:

```
public class Myvalidator:AbstractValidator<Foo> {
  public MyValidator() {
    RuleSet("clientside", () => {
      RuleFor(x => x.Whatever).NotNull();
    });
    
    RuleFor(x=>x...) //other rules
  }
} 
```

In your controller action, you can use the RulesetForClientsideMessages attribute to decorate your action method, which you can use to indicate which ruleset should be used for generating client-side rules.

**Comment by martintro** [comment]
Thanks. I will try that.

Do you know where in the code the rules with conditions are filtered out from client side validation?

**Comment by JeremySkinner** [comment]
They're not filtered out- there's simply no matching clientside adapter for them. The ModelValidatorProvider is what picks the appropriate clientside adaptor for a particular serverside validator:

https://github.com/JeremySkinner/FluentValidation/blob/master/src/FluentValidation.Mvc4/FluentValidationModelValidatorProvider.cs

You could potentially register a new adapter that acts on any DelegatingValidator, unwraps it and calls back into the provider to find a matching adaptor based on the underlying validator

**Comment by martintro** [comment]
Alright. That was the solution I thought would be easiest. Maybe the ruleset is a more cleaner way to solve it though?

**Comment by JeremySkinner** [comment]
I think either approach will work...the ruleset approach means you can do this on a per-validator basis, while the adaptor approach means it applies to all conditional validators which may or may not be what you want. But the downside of the ruleset approach is it makes your validator definitions much more messy.

**Comment by martintro** [comment]
I tried now but I haven't found any solution to access the validatorFactories-member in FluentValidationModelValidatorProvider from my adaptor. Do you have any idea how I find the matchning adaptor from the inner validator?

**Comment by martintro** [comment]
Any ideas?

**Comment by JeremySkinner** [comment]
I could add a protected property to allow you to access the validactorFactories in a derived class. Would that work for you?

**Comment by martintro** [comment]
Then I could just override the provider and the best would be to override the GetModelValidator-method instead of creating an adaptor?

I'm using nuget package FluentValidation.MVC4 with version 5.1.0.0 though and in that version the method is not virtual. Would these two modifications be able to do? If so, how soon could the new version be published?

**Comment by JeremySkinner** [comment]
It'd be in the FluentValidation.MVC5 package....the MVC4 one isn't updated anymore I'm afraid. Earliest I could get a new version out is probably going to be september (I'm away from next week for all of August)

**Comment by martintro** [comment]
Ah, then I think I have to stick with the other option with rulesets. If you don't see a solution without creating a new version?

**Comment by JeremySkinner** [comment]
The only other option is to fork the code for the FluentValidationModelValidatorProvider and keep it in your own project....it's not a particularly large or complicated class.

**Comment by markusgud** [comment]
I have a similar problem where I want to validate a Value member based on an IsRequired member on the same class (named TextField).

Would it be possible to assign a Validator to each instance of a class? Would be cleanest solution if possible.

Or would it be possible to set validators (using the attributes) on derived classes (e.g. DefaultTextField and RequiredTextField). Not very clean though if there are multiple combinations of conditions based on variables like IsRequired.

**Comment by Basek** [comment]
I have simular issue with When. For exemple, I have next rule chain

this.RuleFor(x => x.DateEvent)
                .Cascade(CascadeMode.StopOnFirstFailure)
                .NotEmpty()
                .Must(ValidationHelper.BeAValidDate)
                .When(x => x.CanEdit);

When condition is appling only for Must but client side attributes don't generating and for NotEmpty. If divide it two rule set: one with NotEmpty and other with Must than all work properly, but there are two error message for field and server doing useless work

**Comment by JeremySkinner** [comment]
When applies to *all* validators in the same RuleFor chain, unless you specify ApplyConditionTo.CurrentValidator as the second parameter, so this rule is not compatible with clientside.

**Comment by arfreitas12** [comment]
JeremySkinner,

I have worked on an enhancement to fluent validation that does something similar to what this user is requesting. I too had an issue where I needed to have my validation run on the client only when visible and then not on the server side. I would love to chat with you about my solution at some point and talk about the possibility of adding it into the package.

Regards,
Anthony Freitas
